_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,228 @@
+
+```xml
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ <url>
+  <loc>https://www.artsy.net/show/guy-hepner-kate-moss-exhibiton</loc>
+ </url>
+ ...
+</urlset>
+```
+
+### The Monolith's Rake Task
+
+When Artsy was a Ruby on Rails monolith we generated and wrote static sitemap files to S3 in a Rake task. We iterated over all our data nightly, created a set of URLs and wrote an XML file. We got fairly clever with a `Sitemappable` module that was included in anything that yielded a sitemap.
+
+```ruby
+module Sitemappable
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def sitemappable_priority
+      0.5
+    end
+  end
+end
+```
+
+We also used HAML to generate sitemaps.
+
+```haml
+!!! XML
+%urlset{ xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9" }
+  - documents.each do |document|
+      - type.sitemappable_urls(document, self).each do |url|
+        %url
+          %loc=h url
+```
+
+Of course, there were many more implementation details, but adding a sitemap was easy as we made Mongoid models use `Sitemappable`.
+
+Generating sitemaps took a couple of hours and ran nightly.
+
+### The Dynamic Front-End
+
+By 2015 our front-end was fully decoupled from the monolith that now served as the core API service. The sitemap code still lived there and needed to know about an increasing number of URLs, which seemed totally backwards. We wanted an easier way to maintain and update sitemaps and concluded that by generating sitemaps on the fly we would also see fewer 404 errors during crawling as a side effect. Furthermore, sitemaps increasingly needed information from other places, as we tore the monolith apart into micro-services.
+
+Our solution for dynamic sitemaps queried the API for the number of total items in every collection, then generated a sitemap index with an entry for each page of results. A subsequent query from Google for a given page would query the API for the actual data that was needed to produce the requested sitemap. You can see the entire code [in our open-source artsy.net project](https://github.com/artsy/force/tree/2059f8f5e7faa8213221480c1c9cd3c62e9d5ffb/desktop/apps/sitemaps).
+
+This worked for a while, but introduced new challenges. The first was that we needed individual sitemaps to contain thousands of items or have tens of thousands of sitemaps. That meant tens of megabytes of data returned from either multiple API calls or from a very slow single API query. That problem was even more serious for models that required joining data, and was solved by our first attempt at an orchestration layer that would crawl our own API on a schedule. Today we use GraphQL for that purpose, but we never deprecated the system we put in place during this phase of the sitemap rewrite. That system was barely maintained and riddled with its own issues, starting by the fact that the data crawl began timing out as the amount of data increased sharply.
```

Comment:
`we use GraphQL` could link to some of our past articles or repos on that topic.

I feel the pain described here, but maybe it's not as interesting as just "became unsustainable as our data grew."